### ISSUE TITLE
compute fisher matrix after train?

### ISSUE THREAD

**Issue by jeong-tae** [stokesj/EWC#1]
I am not sure i understand correctly.

it looks like you are compute fisher matrix for current task with samples after the training a current task. am i right?

Then, when moving to third task... does fisher diagonal compute for all previous tasks? or just compute previous task only? I think if the former, EWC is useless because we do not want to see previous. right?

i also implement the EWC using tensorflow. but, i am not sure that i am right. so i am referring your code :D. Thank you for your advanced work!

p.s why do we need to compute fisher diagonal per example? why not per batch?

**Comment by stokesj** [comment]
Yes, I am just computing the Fisher information for the trained neural network using the previous task's data only.

Unfortunately, TensorFlow does not expose the unaggregated gradients, which are required to compute the Fisher information. The workaround I chose was to hardcode the unaggregated gradients directly into the computation graph, using mini batches of fixed size 100 to allow parallelization (computing full-batch directly runs out of memory). It is then a simple matter to obtain the full-batch Fisher information by accumulating 550 mini-batches (see update_fisher_full_batch method). The simpler solution (which prevents parallelization) is to loop over the training data (55000 examples) using minibatches of size 1.

**Comment by jeong-tae** [comment]
what is the difference with calculating a mean of fisher information on aggregated gradients?
The code is summing up and divide by the number of inputs, which seems the value of mean.

unaggregating the gradients make code more complicated, isn't it? why don't you use just mean of aggregated gradients?
